src/librustc_trans/trans/glue.rs

```diff
@@ -384,15 +338,37 @@ fn trans_struct_drop<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
     // discriminant (if any) in case of variant swap in drop code.
     bcx.fcx.schedule_drop_adt_contents(cleanup::CustomScope(contents_scope), v0, t);
 
-    let glue_type = get_drop_glue_type(bcx.ccx(), t);
-    let dtor_ty = bcx.tcx().mk_ctor_fn(class_did, &[glue_type], bcx.tcx().mk_nil());
-    let (_, bcx) = if type_is_sized(bcx.tcx(), t) {
-        invoke(bcx, dtor_addr, &[v0], dtor_ty, DebugLoc::None)
+    let (sized_args, unsized_args);
+    let args: &[ValueRef] = if type_is_sized(tcx, t) {
+        sized_args = [v0];
+        &sized_args
     } else {
-        let args = [Load(bcx, expr::get_dataptr(bcx, v0)), Load(bcx, expr::get_meta(bcx, v0))];
-        invoke(bcx, dtor_addr, &args, dtor_ty, DebugLoc::None)
+        unsized_args = [Load(bcx, expr::get_dataptr(bcx, v0)), Load(bcx, expr::get_meta(bcx, v0))];
+        &unsized_args
     };
 
+    bcx = callee::trans_call_inner(bcx, DebugLoc::None, |bcx, _| {
+        let trait_ref = ty::Binder(ty::TraitRef {
+            def_id: tcx.lang_items.drop_trait().unwrap(),
+            substs: tcx.mk_substs(Substs::trans_empty().with_self_ty(t))
+        });
+        let vtbl = match fulfill_obligation(bcx.ccx(), DUMMY_SP, trait_ref) {
+            traits::VtableImpl(data) => data,
+            _ => tcx.sess.bug(&format!("dtor for {:?} is not an impl???", t))
+        };
+        let dtor_did = tcx.destructor_for_type.borrow()[&def.did];
```

Comment:
Is there a need `destructor_for_type`? The `impl` item should be easy to grab from here. Maybe all of this can be with a single call to a method in `trans::meth`.